src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,291 @@
+        {
+            if (lastUsageStatementIndex == -1)
+            {
+                return default;
+            }
+
+            return new SyntaxList<TStatementSyntax>(parentList
+                .Take(lastUsageStatementIndex + 1)
+                .Skip(declarationStatementIndex + 1));
+        }
+
+        private static TStatementSyntax FindSiblingStatementContainingLastUsage(
+            TStatementSyntax declarationSyntax,
+            ILocalSymbol localVariable,
+            SemanticModel semanticModel,
+            ISyntaxFactsService syntaxFactsService,
+            CancellationToken cancellationToken)
+        {
+            foreach (var nodeOrToken in declarationSyntax.Parent.ChildNodesAndTokens().Reverse())
+            {
+                var node = (TStatementSyntax)nodeOrToken.AsNode();
+                if (node is null)
+                {
+                    continue;
+                }
+
+                if (node == declarationSyntax)
+                {
+                    break; // Ignore the declaration and usages prior to the declaration
+                }
+
+                if (ContainsReference(node, localVariable, semanticModel, syntaxFactsService, cancellationToken))
+                {
+                    return node;
+                }
+            }
+
+            return null;
+        }
+
+        private static bool ContainsReference(
+            SyntaxNode node,
+            ILocalSymbol localVariable,
+            SemanticModel semanticModel,
+            ISyntaxFactsService syntaxFactsService,
+            CancellationToken cancellationToken)
+        {
+            if (syntaxFactsService.IsIdentifierName(node))
+            {
+                var identifierName = syntaxFactsService.GetIdentifierOfSimpleName(node).ValueText;
+
+                return syntaxFactsService.StringComparer.Equals(localVariable.Name, identifierName) &&
+                    localVariable.Equals(semanticModel.GetSymbolInfo(node).Symbol);
+            }
+
+            foreach (var nodeOrToken in node.ChildNodesAndTokens())
+            {
+                var childNode = nodeOrToken.AsNode();
+                if (childNode is null)
+                {
+                    continue;
+                }
+
+                if (ContainsReference(childNode, localVariable, semanticModel, syntaxFactsService, cancellationToken))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static SyntaxList<T> ReplaceRange<T>(SyntaxList<T> syntaxList, int index, int count, IEnumerable<T> newNodes)
+            where T : SyntaxNode
+        {
+            var list = syntaxList.ToList();
+
+            list.RemoveRange(index, count);
+            list.InsertRange(index, newNodes);
+
+            return new SyntaxList<T>(list);
```

Comment:
i don't really see the benefit of this when we have SyntaxEditor which (to me) is a lot cleaner than all the range manipulation stuff.

Reply:
(see https://github.com/dotnet/roslyn/pull/30106#discussion_r223223425)